keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,672 @@
+
+The combined value of `IntialDelaySeconds` and `InitialOffsetDelayMilliseconds` must be greater than or equal to zero ms.
+
+### Existing use of Probe struct fields.
+
+#### InitialDelaySeconds
+https://github.com/kubernetes/kubernetes/blob/7c2e6125694e1aadc78a5fed1cf696872af50a5e/pkg/kubelet/prober/worker.go#L246-L249
+```go
+	// Probe disabled for InitialDelaySeconds.
+	if int32(time.Since(c.State.Running.StartedAt.Time).Seconds()) < w.spec.InitialDelaySeconds {
+		return true
+	}
+```
+
+#### TimeoutSeconds
+https://github.com/kubernetes/kubernetes/blob/7c2e6125694e1aadc78a5fed1cf696872af50a5e/pkg/kubelet/prober/prober.go#L160-L213
+```go
+	timeout := time.Duration(p.TimeoutSeconds) * time.Second
+```
+
+#### ProbeSeconds
+https://github.com/kubernetes/kubernetes/blob/7c2e6125694e1aadc78a5fed1cf696872af50a5e/pkg/kubelet/prober/worker.go#L131-L167
+```go
+func (w *worker) run() {
+	probeTickerPeriod := time.Duration(w.spec.PeriodSeconds) * time.Second
+
+	// If kubelet restarted the probes could be started in rapid succession.
+	// Let the worker wait for a random portion of tickerPeriod before probing.
+	// Do it only if the kubelet has started recently.
+	if probeTickerPeriod > time.Since(w.probeManager.start) {
+		time.Sleep(time.Duration(rand.Float64() * float64(probeTickerPeriod)))
+	}
+
+	probeTicker := time.NewTicker(probeTickerPeriod)
+
+	defer func() {
+		// Clean up.
+		probeTicker.Stop()
+		if !w.containerID.IsEmpty() {
+			w.resultsManager.Remove(w.containerID)
+		}
+
+		w.probeManager.removeWorker(w.pod.UID, w.container.Name, w.probeType)
+		ProberResults.Delete(w.proberResultsSuccessfulMetricLabels)
+		ProberResults.Delete(w.proberResultsFailedMetricLabels)
+		ProberResults.Delete(w.proberResultsUnknownMetricLabels)
+	}()
+
+probeLoop:
+	for w.doProbe() {
+		// Wait for next probe tick.
+		select {
+		case <-w.stopCh:
+			break probeLoop
+		case <-probeTicker.C:
+		case <-w.manualTriggerCh:
+			// continue
+		}
+	}
+}
+```
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+Unit tests, end-to-end tests, and performance tests/benchmarks will need to be added.
+For actual tests to be added, see [the draft implementation](https://github.com/kubernetes/kubernetes/pull/107958/commits)
+
+##### Unit tests
+
+Will add unit tests for new code, specifically in:
+- `pkg/api/pod`: `2022-10-03` - `70.1`
+- `pkg/apis/core/validation`: `2022-10-03` - `82.3`
+- `pkg/kubelet/prober`: `2022-10-03` - `80.7`
+
+##### Integration tests
+
+Not adding integration tests (will be covered by e2e). 
+
+##### e2e tests
+
+Will add e2e tests to `test/e2e/common/node/container_probe.go` to (a) run a sub-second probe, and (b) ensure the sub-second probe completes faster than a standard (1s) probe.
```

Comment:
We have to add performance/scale tests for this feature, since TCP/HTTP probes can exhaust quickly all the conntrack or ephemeral ports available in the node https://github.com/kubernetes/kubernetes/issues/89898#issuecomment-1383207322

The number of probes is equal to PrxCxP
- Pr: probe per containers that can be readines, liveness and/or startup
- C: containers per Pod
- P: number of pods

Assume a worst case of 100 pods with 2 containers each and 3 probes per container configured with 1 seconds
100x2x3 x 1 req/sec = 600 req per second.

To free the socket the OS has to wait for the TIME_WAIT period (default 60sec), and that will free the port and the conntrack entry.

These network resources are governed by these options
```
cat /proc/sys/net/ipv4/ip_local_port_range
32768   60999 

net.netfilter.nf_conntrack_max = 262144
```

This means that a rate of 600 request per second (we are not going to free resources until 60 second later), we'll be holding 600x59 = 35400 ephemeral ports and we only have 28321 per default.

This means that kubelet can not create any new connection and we are still not using subsecond.

I have plans to alleviate this problem https://github.com/kubernetes/kubernetes/issues/89898#issuecomment-1384673191 , but I really think that this KEP should take this into account add e2e test should be part of this, maybe not for alpha but I really think for beta we should require performance/scale validation for something that operates at the subsecond level

cc: @wojtek-t @johnbelamaric

Reply:
I showed my ignorance about probes, it seems readiness and start probes have some interesting behavior, but besides that the network part are ok.

The good thing is that I have a PR to fix it and I have a test how probes behave at scale, that can be modified to test this new feature
https://github.com/kubernetes/kubernetes/pull/115143/commits/44f9c521abcbcca84a3d009d65a276c54164a4c6

Reply:
ok, one thing I need to verify because someone with more knowledge than me in this area told me, is that the ephemeral port range may depend only on the 5-tuple, so different pods should be able to reuse the same ephemeral port for probes? that will make this problem less critical, still stands the one that are a considerable amount of resources used